crates/ruff/src/logging.rs

```diff
@@ -1,8 +1,32 @@
+use std::sync::Mutex;
+
 use anyhow::Result;
 use colored::Colorize;
 use fern;
 use log::Level;
+use once_cell::sync::Lazy;
+use rustc_hash::FxHashSet;
+
+pub(crate) static WARNINGS: Lazy<Mutex<FxHashSet<String>>> = Lazy::new(Mutex::default);
```

Comment:
Using a `String` as the key has the potential risk that two keys end up with the same name. 

We can use `TypeId` to avoid this. I've done this before for the [`Label`](https://github.com/charliermarsh/ruff/blob/e356ebcc6b6984a2e02dfe55f97f5d36ce0574a5/crates/ruff_formatter/src/format_element/tag.rs#L243) struct in the formatter. 

```rust
#[derive(Eq, PartialEq, Copy, Clone, Debug)]
pub struct WarningId {
    id: TypeId,
	// Pretty name for debugging
    #[cfg(debug_assertions)]
    label: &'static str,
}

impl WarningId {
    pub fn of<T: ?Sized + 'static>() -> Self {
        Self {
            id: TypeId::of::<T>(),
            #[cfg(debug_assertions)]
            label: type_name::<T>(),
        }
    }
}
```

Usage: 

```rust
struct MyWarning;

let id = WarningId::of::<MyWarning>("Some name");
```


Using `TypeId` further has the advantage that using a simple `Vec` and scanning it is probably cheaper (this is based on the assumption that users will have few warnings, let's say less than 16).

Reply:
Hm, okay, you indeed want to match on a string here. The above approach could still work but may be more complicated. 

I would still argue that using a `Vec` instead of a `FxHashMap` is probably cheaper since the lookups are rare and the table only contains few entries.